lib/realtime_web/channels/realtime_channel.ex

```diff
@@ -691,4 +550,129 @@ defmodule RealtimeWeb.RealtimeChannel do
       }
     )
   end
+
+  defp is_new_api(%{"config" => _}), do: true
+  defp is_new_api(_), do: false
+
+  defp pg_change_params(true, params, channel_pid, claims, _) do
+    send(self(), :sync_presence)
+
+    case get_in(params, ["config", "postgres_changes"]) do
+      [_ | _] = params_list ->
+        Enum.map(params_list, fn params ->
+          %{
+            id: UUID.uuid1(),
+            channel_pid: channel_pid,
+            claims: claims,
+            params: params
+          }
+        end)
+
+      _ ->
+        []
+    end
+  end
+
+  defp pg_change_params(false, params, channel_pid, claims, sub_topic) do
+    case String.split(sub_topic, ":", parts: 3) do
+      [schema, table, filter] -> %{"schema" => schema, "table" => table, "filter" => filter}
+      [schema, table] -> %{"schema" => schema, "table" => table}
+      [schema] -> %{"schema" => schema}
+    end
+
+    [
+      %{
+        id: UUID.uuid1(),
+        channel_pid: channel_pid,
+        claims: claims,
+        params: params
+      }
+    ]
+  end
+
+  defp postgres_cdc_subscribe(%{pg_change_params: []}), do: []
+
+  defp postgres_cdc_subscribe(opts) do
+    %{
+      is_new_api: is_new_api,
+      pg_change_params: pg_change_params,
+      transport_pid: transport_pid,
+      serializer: serializer,
+      topic: topic,
+      tenant: tenant,
+      module: module
+    } = opts
+
+    ids =
+      Enum.map(pg_change_params, fn %{id: id, params: params} ->
+        {UUID.string_to_binary!(id), :erlang.phash2(params)}
+      end)
+
+    subscription_metadata =
+      {:subscriber_fastlane, transport_pid, serializer, ids, topic, tenant, is_new_api}
+
+    metadata = [metadata: subscription_metadata]
+
+    PostgresCdc.subscribe(module, pg_change_params, tenant, metadata)
+
+    send(self(), :postgres_subscribe)
+
+    pg_change_params
+  end
+
+  defp add_id_to_postgres_changes(pg_change_params) do
+    Enum.map(pg_change_params, fn %{params: params} ->
+      id = :erlang.phash2(params)
+      Map.put(params, :id, id)
+    end)
+  end
+
+  defp check_tenant_connection(extensions) do
+    extensions
+    |> Enum.map(fn %{settings: settings} ->
+      ssl_enforced = Helpers.default_ssl_param(settings)
+
+      host = settings["db_host"]
+      port = settings["db_port"]
+      name = settings["db_name"]
+      user = settings["db_user"]
+      password = settings["db_password"]
+      socket_opts = settings["db_socket_opts"]
+
+      opts = %{
+        host: host,
+        port: port,
+        name: name,
+        user: user,
+        pass: password,
+        socket_opts: socket_opts,
+        pool: 1,
+        queue_target: 1000,
+        ssl_enforced: ssl_enforced
+      }
+
+      with {:ok, conn} <- Helpers.connect_db(opts),
```

Comment:
We really can't do this here because every client is going to try and fire up it's own connection to the database when they attempt to join. 

We need a single process on the whole cluster which is started when the first client joins, and remains connected such that when subsequent clients join we know the database is healthy and can do a quick check of `:syn`.

This is basically the way the PostgresCdcRls and PostgresStream work. There is one database connection (tree) on the whole cluster which powers all the clients.

Reply:
The reason we don't just use these is because a client may not want to stream changes from the database.

Reply:
ok will change things around to fit that, wanted to avoid a more complex solution but you are right an influx of 5k users trying to do this would kill everything 🤦 

will refactor this to fit that model with the following:
* One Genserver that on start checks the status of the tenant and keeps updating the status with `:syn`
* That Genserver should be started on the node nearest to the tenant database and identified by the tenant external id

Reply:
Realtime initiates multiple connection pools to the tenant's db and I don't believe this check ensures that all of them will connect successfully. For example, the database could reach its maximum connection limit shortly after this check 🤷

Btw, it seems currently that `conn` is terminated only if the connection is successful. Also, it's better to use `GenServer.stop` for stopping, cos `conn` is an instance of the GenServer behavior

Reply:
lol this discussion will probably never end. I'm in the one-line-per-alias camp but at this point I don't particularly care. Should just setup a credo rule or something per project and just do what it says. @abc3 you can be the decider for Realtime :D